runtime/gc_vlhgc/CopyForwardCompactGroup.hpp

```diff
@@ -132,6 +139,91 @@ class MM_CopyForwardCompactGroup
+				_TLHRemainderBase = newAlignedTLHRemainderBase;
+				_TLHRemainderTop = newAlignedTLHRemainderTop;
+				aligned = true;
+			}
+			/* else aligned TLHRemainder is smaller than tlhSurvivorDiscardThreshold, need to be reset */
+		}
+		return aligned;
+	}
+
+	MMINLINE void abandonTLHRemainder(MM_EnvironmentVLHGC* env, uintptr_t compactGroup, bool preserveRemainder)
+	{
+		if (NULL != _TLHRemainderBase) {
+			Assert_MM_true(NULL != _TLHRemainderTop);
+			/* make it a walkable hole */
+			env->_cycleState->_activeSubSpace->abandonHeapChunk(_TLHRemainderBase, _TLHRemainderTop);
+			MM_GCExtensions* extensions =MM_GCExtensions::getExtensions(env);
+			MM_HeapRegionDescriptorVLHGC *region = (MM_HeapRegionDescriptorVLHGC*)extensions->heapRegionManager->tableDescriptorForAddress(_TLHRemainderBase);
+
+			/* only preserve for non Nursery regions */
+			if (!preserveRemainder || (MM_CompactGroupManager::getRegionAgeFromGroup(env, compactGroup) < extensions->tarokNurseryMaxAge._valueSpecified) || !region->containsObjects()) {
+				/* discard the remainder */
+				discardHeapChunk(_TLHRemainderBase, _TLHRemainderTop, region);
+				resetTLHRemainder();
+			}
+		} else {
+			Assert_MM_true(NULL == _TLHRemainderTop);
+		}
+	}
+
+	MMINLINE void discardHeapChunk(void *base, void *top, MM_HeapRegionDescriptorVLHGC *region)
+	{
+		uintptr_t discardSize = ((uintptr_t)top - (uintptr_t)base);
+		_discardedBytes += discardSize;
+		region->getMemoryPool()->incrementDarkMatterBytes(discardSize);
+	}
+
+	/**
+	 * check and prepare TLHRemainder ready for using by collector
+	 *  1, reset TLHRemainder if it is invalid.
+	 *  2, reset TLHRemainder if it is in collection set
+	 *  3, align TLHRemainder with Card
+     * @return region, if remainders are non-null, and regions is not already converted to survivor
+     * @return NULL, otherwise
+	 */
+	MMINLINE MM_HeapRegionDescriptorVLHGC* prepareTLHRemainder(MM_EnvironmentVLHGC* env)
+	{
+		MM_HeapRegionDescriptorVLHGC *region = NULL;
+		if (NULL != _TLHRemainderBase) {
+			/* first PGC after global or GMP need to reset TLHRemainder */
+			bool needResetTLHRemainder = !static_cast<MM_CycleStateVLHGC*>(env->_cycleState)->_schedulingDelegate->areCopyCacheRemaindersValid();
+			MM_GCExtensions* extensions =MM_GCExtensions::getExtensions(env);
+			if (!needResetTLHRemainder) {
+				region = (MM_HeapRegionDescriptorVLHGC*)extensions->heapRegionManager->tableDescriptorForAddress(_TLHRemainderBase);
+				if (region->_markData._shouldMark || region->isFreeOrIdle()) {
+					/* the region is part of collection set, reset TLHRemainder related with the region */
+					needResetTLHRemainder = true;
+				} else {
+					/* align TLHRemainder with card */
+					needResetTLHRemainder = !alignTLHRemainder(extensions->tlhSurvivorDiscardThreshold);
+				}
+			}
+
+			if (needResetTLHRemainder) {
+				resetTLHRemainder();
+				region = NULL;
+			}
+			if ((NULL != region) && region->isSurvivorRegion()) {
```

Comment:
minor.... I believe these 2 _ifs_ could be connected (if needResetTLHRemainder is false, region should be non-zero), although not sure if helps with readablity of the code - up to you, if you want to change

 ```
			if (needResetTLHRemainder) {
				resetTLHRemainder();
				region = NULL;
			} else if (region->isSurvivorRegion()) {
				region = NULL;
			}
```